sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.spill
+
+import java.io._
+import java.nio.ByteBuffer
+import java.nio.channels.{Channels, FileChannel, WritableByteChannel}
+import java.nio.file.StandardOpenOption
+import java.util
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.mutable
+
+import ai.rapids.cudf._
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuColumnVectorFromBuffer, GpuCompressedColumnVector, GpuDeviceManager, HostAlloc, HostMemoryOutputStream, MemoryBufferToHostByteBufferIterator, RapidsConf, RapidsHostColumnVector}
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.RapidsPluginImplicits.AutoCloseableSeq
+import com.nvidia.spark.rapids.format.TableMeta
+import com.nvidia.spark.rapids.internal.HostByteBufferIterator
+import org.apache.commons.io.IOUtils
+
+import org.apache.spark.{SparkConf, SparkEnv, TaskContext}
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.rapids.{GpuTaskMetrics, RapidsDiskBlockManager}
+import org.apache.spark.sql.rapids.execution.SerializedHostTableUtils
+import org.apache.spark.sql.rapids.storage.RapidsStorageUtils
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.storage.BlockId
+
+/**
+ * Spark-RAPIDS Spill Framework
+ *
+ * The spill framework tracks device/host/disk object lifecycle in the RAPIDS Accelerator
+ * for Apache Spark. A set of stores is used to track these objects, which are wrapped in
+ * "handles" that describe the state of each to the user and to the framework.
+ *
+ * This file comment covers some pieces of the framework that are worth knowing up front.
+ *
+ * Ownership:
+ *
+ * Any object handed to the framework via the factory methods for each of the handles
+ * should not be used directly by the user. The framework takes ownership of all objects.
+ * To get a reference back, call the `materialize` method, and always close what the framework
+ * returns.
+ *
+ * CUDA/Host synchronization:
+ *
+ * We assume all device backed handles are completely materialized on the device (before adding
```

Comment:
I added info in this comment. We could, at least for the GPU handles, add an event per handle and we could record on that event, instead of synchronizing. I think this means I should move the synchronization to the framework, right now (as you spotted), I let callers synchronize before they call the factory methods.